crates/biome_js_analyze/src/lint/nursery/no_magic_numbers.rs

```diff
@@ -0,0 +1,239 @@
+};
+use biome_console::markup;
+use biome_deserialize_macros::Deserializable;
+use biome_js_syntax::{
+    AnyJsExpression, JsAssignmentExpression, JsCallExpression, JsSyntaxKind, JsUnaryOperator,
+    JsVariableDeclarator,
+};
+use biome_rowan::{AstNode, SyntaxNode, TextRange};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+const MAX_ARRAY_LENGTH: i64 = (2_i64.pow(32)) - 1;
+
+#[derive(
+    Debug, Default, Clone, Serialize, Deserialize, Deserializable, PartialEq, Eq, JsonSchema,
+)]
+pub struct NoMagicNumbersConfig {
+    detect_objects: bool,
+    enforce_const: bool,
+    ignore: Vec<i64>,
+    ignore_array_indexes: bool,
+    ignore_default_values: bool,
+    ignore_class_field_initial_values: bool,
+}
+
+declare_lint_rule! {
+    /// Disallow magic numbers
+    ///
+    /// This rule aims to make code more maintainable by ensuring that special numbers
+    /// are declared as constants with meaningful names.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// let total = price * 1.23; // Magic number for tax rate
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const TAX_RATE = 1.23;
+    /// let total = price * TAX_RATE;
+    /// ```
+    pub NoMagicNumbers {
+        version: "next",
+        name: "noMagicNumbers",
+        language: "js",
+        sources: &[RuleSource::Eslint("no-magic-numbers")],
+        recommended: false,
+    }
+}
+
+#[derive(Debug)]
+pub struct NumberContext {
+    range: TextRange,
+    raw: String,
+    is_const_violation: bool,
+}
+
+impl Rule for NoMagicNumbers {
+    type Query = Ast<AnyJsExpression>;
+    type State = NumberContext;
+    type Signals = Option<Self::State>;
+    type Options = NoMagicNumbersConfig;
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let config = ctx.options();
+
+        match node {
+            AnyJsExpression::AnyJsLiteralExpression(literal_expr) => {
+                if let Some(literal) = literal_expr.as_js_number_literal_expression() {
+                    let value_token = literal.value_token().ok()?;
+                    let value_text = value_token.text();
+                    let value = value_text.parse::<i64>().ok()?;
+                    if is_allowed_number(literal.syntax(), value, config) {
+                        return None;
+                    }
+
+                    let is_const_violation = check_const_violation(literal.syntax(), config);
+
+                    Some(NumberContext {
+                        range: literal.range(),
+                        raw: value_token.text().to_string(),
+                        is_const_violation,
+                    })
+                } else {
+                    return None;
+                }
+            }
+            AnyJsExpression::JsUnaryExpression(unary) => {
+                if let Ok(operator) = unary.operator() {
```

Comment:
Here's a thing about our APIs. As you might have noticed, the majority of methods return a `Result`. That's because Biome uses a CST, which means that any possible node might have syntax errors. 

Now, inside lint rules, we don't care about broken syntax, this means that you can safely use `.ok()?` for each method that returns a `Result`. In this case, all `Ok()` can be removed in favour of `.ok()?`.

The same can be applied applied, more or less to the `.as_*` methods. If you're interested to only **one variant**, you can the `as_` method, which means you aren't interested on the other ones. So use the try operator `?`.  This will simplify the code.